onnxruntime/core/framework/provider_adapter.h

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/session/onnxruntime_c_api_ep.h"
+#include "core/framework/compute_capability.h"
+#include "core/framework/error_code_helper.h"
+#include "core/framework/kernel_registry.h"
+#include "core/session/allocator_adapters.h"
+
+namespace onnxruntime {
+
+class DataTransferAdapter : public IDataTransfer {
+public:
+  DataTransferAdapter(OrtExecutionProvider* ep) : ep_impl_(ep) {}
+  virtual bool CanCopy(const OrtDevice& src_device, const OrtDevice& dst_device) const override {
+    return ep_impl_->CanCopy(&src_device, &dst_device);
+  }
+
+  virtual common::Status CopyTensor(const Tensor& src, Tensor& dst) const override {
+    OrtMemoryInfoDeviceType source_device_type = static_cast<OrtMemoryInfoDeviceType>(src.Location().device.Type());
+    OrtMemoryInfoDeviceType target_device_type = static_cast<OrtMemoryInfoDeviceType>(dst.Location().device.Type());
+    OrtMemoryType source_mem_type = static_cast<OrtMemoryType>(src.Location().device.MemType());
+    return ToStatus(ep_impl_->CopyTensor(src.DataRaw(), source_device_type, source_mem_type, dst.MutableDataRaw(), target_device_type, src.SizeInBytes(), nullptr));
+  }
+
+  virtual common::Status CopyTensorAsync(const Tensor& src, Tensor& dst, Stream& stream) const override {
+    OrtMemoryInfoDeviceType source_device_type = static_cast<OrtMemoryInfoDeviceType>(src.Location().device.Type());
+    OrtMemoryInfoDeviceType target_device_type = static_cast<OrtMemoryInfoDeviceType>(dst.Location().device.Type());
+    OrtMemoryType source_mem_type = static_cast<OrtMemoryType>(src.Location().device.MemType());
+    return ToStatus(ep_impl_->CopyTensor(src.DataRaw(), source_device_type, source_mem_type, dst.MutableDataRaw(), target_device_type, src.SizeInBytes(), stream.GetHandle()));
+  }
+private:
+  OrtExecutionProvider* ep_impl_;
+};
+
+class ExecutionProviderAdapter : public IExecutionProvider {
+public:
+  ExecutionProviderAdapter(OrtExecutionProvider* ep) : IExecutionProvider(ep->type, ep->default_device ? *(ep->default_device) : OrtDevice()), ep_impl_(ep) {
+    builtin_ep_ = false;
+    if (ep_impl_->RegisterKernels) {
+      kernel_registry_ = std::make_shared<KernelRegistry>();
+      ep_impl_->RegisterKernels(reinterpret_cast<OrtKernelRegistry*>(kernel_registry_.get()));
+    }
+  }
+  virtual std::vector<std::unique_ptr<ComputeCapability>> GetCapability(const GraphViewer& graph_viewer, const IKernelLookup& kernel_lookup) const override {
+    size_t cnt = 0;
+    OrtIndexedSubGraph** indexed_subgraph = nullptr;
+    if (ep_impl_->GetCapability) ep_impl_->GetCapability(ep_impl_, reinterpret_cast<const OrtGraphViewer*>(&graph_viewer), &cnt, &indexed_subgraph);
```

Comment:
This is currently passing an `OrtIndexedSubGraph***` to the EP's GetCapability() function and asking the EP to allocate an array of pointers to OrtIndexSubGraph objects. Because the EP is allocating the memory, we currently have a separate API to allow the EP to delete this memory.

I wonder if it would be simpler to allow ORT to pass in an `OrtAllocator` to the EP. The EP would use this ORT-owned allocator to allocate memory for the array. This woud remove the need for a separate C API to clean up the memory. Also, it would allow the parameter to be a `OrtIndexedSubGraph**` instead of `OrtIndexedSubGraph***`.